testing/bdd_test.ts

```diff
@@ -0,0 +1,1623 @@
+
+  const baseStepOptions: Omit<Deno.TestStepDefinition, "name" | "fn"> = {
+    ignore: false,
+    sanitizeExit: true,
+    sanitizeOps: true,
+    sanitizeResources: true,
+  };
+
+  const baseOptions: Omit<Deno.TestDefinition, "name" | "fn"> = {
+    ...baseStepOptions,
+    only: false,
+    permissions: "inherit",
+  };
+
+  interface GlobalContext {
+    allTimer: number;
+    eachTimer: number;
+  }
+
+  let timerIdx = 1;
+  const timers = new Map<number, number>();
+  function hookFns() {
+    timerIdx = 1;
+    timers.clear();
+    return {
+      beforeAllFn: spy(async function (this: GlobalContext) {
+        await Promise.resolve();
+        this.allTimer = timerIdx++;
+        timers.set(this.allTimer, setTimeout(() => {}, 10000));
+      }),
+      afterAllFn: spy(async function (this: GlobalContext) {
+        await Promise.resolve();
+        clearTimeout(timers.get(this.allTimer));
+      }),
+      beforeEachFn: spy(async function (this: GlobalContext) {
+        await Promise.resolve();
+        this.eachTimer = timerIdx++;
+        timers.set(this.eachTimer, setTimeout(() => {}, 10000));
+      }),
+      afterEachFn: spy(async function (this: GlobalContext) {
+        await Promise.resolve();
+        clearTimeout(timers.get(this.eachTimer));
+      }),
+    };
+  }
+
+  await t.step("global hooks", async () => {
+    const test = stub(Deno, "test"),
+      fns = [spy(), spy()],
+      { beforeAllFn, afterAllFn, beforeEachFn, afterEachFn } = hookFns();
+
+    try {
+      beforeAll(beforeAllFn);
+      afterAll(afterAllFn);
+
+      beforeEach(beforeEachFn);
+      afterEach(afterEachFn);
+
+      assertEquals(it({ name: "example 1", fn: fns[0] }), undefined);
+      assertEquals(it({ name: "example 2", fn: fns[1] }), undefined);
+
+      assertSpyCalls(fns[0], 0);
+      assertSpyCalls(fns[1], 0);
+
+      assertSpyCall(test, 0);
+      const call = test.calls[0];
+      const options = call.args[0] as Deno.TestDefinition;
+      assertEquals(Object.keys(options).sort(), ["fn", "name"]);
+      assertEquals(options.name, "global");
+
+      const context = new TestContext();
+      const result = options.fn(context);
+      assertStrictEquals(Promise.resolve(result), result);
+      assertEquals(await result, undefined);
+      assertSpyCalls(context.spies.step, 2);
+    } finally {
+      TestSuiteInternal.reset();
+      test.restore();
+    }
+
+    let fn = fns[0];
+    assertSpyCall(fn, 0, {
+      self: { allTimer: 1, eachTimer: 2 },
+      args: [],
+      returned: undefined,
+    });
+    assertSpyCalls(fn, 1);
+
+    fn = fns[1];
+    assertSpyCall(fn, 0, {
+      self: { allTimer: 1, eachTimer: 3 },
+      args: [],
+      returned: undefined,
+    });
+    assertSpyCalls(fn, 1);
+
+    assertSpyCalls(beforeAllFn, 1);
+    assertSpyCalls(afterAllFn, 1);
+    assertSpyCalls(beforeEachFn, 2);
+    assertSpyCalls(afterEachFn, 2);
+  });
+
+  await t.step("it", async (t) => {
+    async function assertOptions<T>(
```

Comment:
This test utility looks very complex. Please add doc comment which describes the purpose of this and what is checked with this.

Reply:
I can do that. These are just used internally. Basically every step in this block of tests have similar assertions regarding Deno.test getting called when you call it, and it verifies Deno.test gets called with the correct options.

Reply:
> These are just used internally.

Yes, but such comments are useful for the maintainers and contributors to better understand the test cases. I'd appreciate if the doc comment is added to these utils, especially with description about what is expected to be passed as `cb` in them